2022-RADIUSS-AWS/JupyterNotebook/radiuss-aws-flux.ipynb

```diff
@@ -0,0 +1,944 @@
+   "execution_count": null,
+   "id": "institutional-vocabulary",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!flux jobs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "77ca4277",
+   "metadata": {},
+   "source": [
+    "Since those jobs won't ever exit (and we didn't specify a timelimit), let's kill them off now and free up the resources."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "46dd8ec8-6c64-4d8d-9a00-949f5f58c07b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!flux job killall -f\n",
+    "!flux jobs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "544aa0a9",
+   "metadata": {},
+   "source": [
+    "We can use the `flux mini batch` command to easily created nested flux instances.  When `flux mini batch` is invoked, Flux will automatically create a nested instance that spans the resources allocated to the job, and then Flux runs the batch script passed to `flux mini batch` on rank 0 of the nested instance. \"Rank\" refers to the rank of the Tree-Based Overlay Network (TBON) used by the Flux brokers: https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man1/flux-broker.html\n",
+    "\n",
+    "While a batch script is expected to launch parallel jobs using `flux mini run` or `flux mini submit` at this level, nothing prevents the script from further batching other sub-batch-jobs using the `flux mini batch` interface, if desired.\n",
+    "\n",
+    "Note: Flux also provides a `flux mini alloc` which is an interactive version of `flux mini batch`, but demonstrating that in a Jupyter notebook is difficult due to the lack of pseudo-terminal."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "blank-carpet",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!flux mini batch --nslots=2 --cores-per-slot=1 --nodes=2 ./sleep_batch.sh\n",
+    "!flux mini batch --nslots=2 --cores-per-slot=1 --nodes=2 ./sleep_batch.sh"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "da98bfa1",
+   "metadata": {},
+   "source": [
+    "The contents of `sleep_batch.sh`:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "pregnant-witch",
+   "metadata": {},
+   "source": [
+    "``` bash \n",
+    "    #!/bin/bash\n",
+    "  \n",
+    "    echo \"Starting my batch job\"\n",
+    "    echo \"Print the resources allocated to this batch job\"\n",
+    "    flux resource list\n",
+    "\n",
+    "    echo \"Use sleep to emulate a parallel program\"\n",
+    "    echo \"Run the program at a total of 2 processes each requiring\"\n",
+    "    echo \"1 core. These processes are equally spread across 2 nodes.\"\n",
+    "    flux mini run -N 2 -n 2 sleep 30\n",
+    "    flux mini run -N 2 -n 2 sleep 30\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "edff8993-3c39-4f46-939d-4c8be5739fbc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!flux jobs\n",
```

Comment:
We should note that jobs that are instances are colored _blue_ in `flux jobs` output.